BaseClasses.py

```diff
@@ -604,38 +612,157 @@ def all_done() -> bool:
+        found = 0
+        for item_name in self._multiworld.worlds[self.player].item_name_groups[item_name_group]:
+            found += self.prog_items[item_name]
+            if found >= count:
+                return True
+        return False
+
+    def count_group(self, item_name_group: str) -> int:
+        found = 0
+        for item_name in self._multiworld.worlds[self.player].item_name_groups[item_name_group]:
+            found += self.prog_items[item_name]
+        return found
+
+    def item_count(self, item: str) -> int:
+        return self.prog_items[item]
+
+    def collect(self, item: Union[Item, Iterable[Item]], event: bool = False,
+                location: Optional[Location] = None) -> List[bool] | bool:
+        if location:
+            self.locations_checked.add(location)
+        if isinstance(item, Iterable):
+            # not sure what this should return but this at least guarantees they all get called :shrug:
+            return [self.collect(item_, event) for item_ in item]
+        changed = self._multiworld.worlds[item.player].collect(self._parent, item)
+        if not changed and event:
+            self.prog_items[item.name] += 1
+            changed = True
+        self.stale = True
+
+        if changed and not event:
+            self.sweep_for_events()
+        return changed
+
+    def remove(self, item: Item) -> None:
+        changed = self._multiworld.worlds[item.player].remove(self._parent, item)
+        if changed:
+            self.reachable_regions = set()
+            self.blocked_connections = set()
+            self.stale = True
+
+    def get(self, key, default) -> Any:
+        return self.prog_items.get(key, default)
+
+    def values(self):
+        return self.prog_items.values()
+
+    def keys(self):
+        return self.prog_items.keys()
+
+    def __iter__(self) -> Iterator[str]:
+        return self.prog_items.__iter__()
+
+    def __getitem__(self, item) -> int:
+        return self.prog_items[item]
+
+    def __setitem__(self, key, value) -> None:
+        return self.prog_items.__setitem__(key, value)
+
+    def __delitem__(self, key) -> None:
+        del self.prog_items[key]
+
+
+class CollectionState:
+    prog_items: Dict[int, PlayerState]  # to be removed
+    states: Dict[int, PlayerState]
     multiworld: MultiWorld
-    reachable_regions: Dict[int, Set[Region]]
-    blocked_connections: Dict[int, Set[Entrance]]
     events: Set[Location]
     path: Dict[Union[Region, Entrance], PathValue]
     locations_checked: Set[Location]
-    stale: Dict[int, bool]
     additional_init_functions: List[Callable[[CollectionState, MultiWorld], None]] = []
     additional_copy_functions: List[Callable[[CollectionState, CollectionState], CollectionState]] = []
 
-    def __init__(self, parent: MultiWorld):
-        self.prog_items = {player: Counter() for player in parent.player_ids}
+    def __init__(self, parent: MultiWorld, player: Optional[int] = None):
```

Comment:
I wonder if it could make sense to have player be an iterable instead?

I can't think of the use case for it off the top of my head but it's trivial to support it so it might be worth including imo